main/src/ca/uwaterloo/flix/language/phase/CodeGen.scala

```diff
@@ -630,11 +630,65 @@ object CodeGen extends Phase[ExecutableAst.Root, ExecutableAst.Root]{
       // Invoking the constructor
       visitor.visitMethodInsn(INVOKESPECIAL, decorate(clazzName), "<init>", s"(${desc})V", false)
 
-    case Expression.Ref(exp, tpe, loc) => ??? // TODO
+    case Expression.Ref(exp, tpe, loc) =>
+      // Adding source line number for debugging
+      addSourceLine(visitor, loc)
+      // Class appropriate reference class
+      val clazz = getReferenceClazz(tpe)
+      // Get internal name of the class
+      val clazzInternalName = asm.Type.getInternalName(clazz)
+      // Get constructor of the class
+      val constructorDescriptor = asm.Type.getConstructorDescriptor(clazz.getConstructors.last)
+      // Create a new reference object
+      visitor.visitTypeInsn(NEW, clazzInternalName)
```

Comment:
There is a tiny problem here.

If I give you the expression `f(1 + 1)` you expect the inner expression to be evaluated first, i.e. `1 + 1` before the call to `f`. 

Similarly, if I give you `f(???)` you would not expect `f` to be called, since `???` crashes.

Now, in this implementation of ref, if you have `ref (1 + 1)` you actually allocate the reference cell first and then you evaluate the inner expression. It would be more correct (but mostly just robust) to evaluate the expression first, and only then allocate the reference cell.